File: 3. Longest Substring Without Repeating Characters.py
class Solution(object):
    # def lengthOfLongestSubstring(self, s):
    #     """
    #     :type s: str
    #     :rtype: int
    #     """
    #     arr = []
    #     result = 0
    #     for i in range(len(s)):
    #         if s[i] in arr:
    #             arr=arr[arr.index(s[i])+1:]
    #         arr.append(s[i])
    #         result = max(len(arr),result)
    #     return result
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        usedChar = dict()
        start=result = 0
        for i,c in enumerate(s):
            if c in usedChar and start<=usedChar[c]:
                start=usedChar[c]+1
            else:
                if i-start+1>result:
                    result = i-start+1
            usedChar[c]=i

        return result

File: 3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
// js 还是这个更快  (⊙﹏⊙)b
// var lengthOfLongestSubstring = function(s) {
//     let subArr=[];
//     let max = 0;
//     for(let i=0,len=s.length;i<len;i++){
//         let index = subArr.indexOf(s[i]);
//         if(index>=0){
//             subArr.splice(0,index+1)
//         }
//         subArr.push(s[i]);
//         max=subArr.length>=max?subArr.length:max;
//     }
//     return max
// };
var lengthOfLongestSubstring = function(s) {
    let used={};
    let start=max = 0;
    for(let i=0,len=s.length;i<len;i++){
        if(typeof used[s[i]] !== 'undefined' && start<=used[s[i]]){
            start=used[s[i]]+1
        }else{
            max=Math.max(i-start+1,max)
        }
        used[s[i]]=i

    }
    return max
};
